rust_src/src/buffers.rs

```diff
@@ -336,43 +354,23 @@ pub fn overlay_buffer(overlay: LispObject) -> LispObject {
 }
 
 pub fn recenter_overlay_lists(mut buf: LispBufferRef, pos: ptrdiff_t) {
-    // let mut overlay: LispObject;
-    let (mut prev, mut tail, mut next): (
-        Option<LispOverlayRef>,
-        LispOverlayRef,
-        Option<LispOverlayRef>,
-    ) = (
-        None,
-        LispOverlayRef::new((*buf).overlays_before as *mut Lisp_Overlay),
-        None,
-    );
-    loop {
-        if tail.as_ptr() as ptrdiff_t == 0 {
-            break;
-        }
-        let (beg, end, curNext) = (tail.start(), tail.end(), tail.next);
-        if end.as_fixnum_coerce_marker_or_error() as ptrdiff_t > pos {
-            let begPtr = beg.as_fixnum_coerce_marker_or_error() as ptrdiff_t;
-
-            match prev {
-                Some(mut tailPtr) => tailPtr.next = curNext,
-                None => set_buffer_overlays_before(buf, curNext),
-            }
-
-            loop {
-                let (mut other_prev, mut other): (
-                    Option<LispOverlayRef>,
-                    LispOverlayRef,
-                ) = (
-                    None,
-                    LispOverlayRef::new((*buf).overlays_after as *mut Lisp_Overlay),
-                );
-            }
-        }
-        // TODO: check loop changes
-        prev = Some(tail);
-        tail = next.unwrap()
+    let before = LispOverlayRef::new((*buf).overlays_before as *mut Lisp_Overlay);
+    let start: (Option<LispOverlayRef>, Option<LispOverlayRef>) = (None, Some(before));
+    let toMove: Option<(Option<LispOverlayRef>, Option<LispOverlayRef>)> = before
+        .iter()
+        .scan(start, |pair, current: LispOverlayRef| {
+            Some((pair.1, Some(current)))
+        })
+        .find(|pair| {
+            (pair.1).unwrap().end().as_fixnum_coerce_marker_or_error() as ptrdiff_t > pos
+        });
```

Comment:
Why is `pair.1` an Option type? it is always a Some value as the unwrap demonstrates.